filebeat/input/tcp/input.go

```diff
@@ -0,0 +1,111 @@
+package tcp
+
+import (
+	"time"
+
+	"github.com/elastic/beats/filebeat/channel"
+	"github.com/elastic/beats/filebeat/harvester"
+	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/filebeat/inputsource/tcp"
+	"github.com/elastic/beats/filebeat/util"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/atomic"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+func init() {
+	err := input.Register("tcp", NewInput)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Input for TCP connection
+type Input struct {
+	server  *tcp.Server
+	started atomic.Bool
+	outlet  channel.Outleter
+	config  *config
+	log     *logp.Logger
+}
+
+// NewInput creates a new TCP input
+func NewInput(
+	cfg *common.Config,
+	outlet channel.Factory,
+	context input.Context,
+) (input.Input, error) {
+	cfgwarn.Experimental("TCP input type is used")
+
+	out, err := outlet(cfg, context.DynamicFields)
+	if err != nil {
+		return nil, err
+	}
+
+	forwarder := harvester.NewForwarder(out)
+
+	config := defaultConfig
+	err = cfg.Unpack(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	cb := func(data []byte, metadata tcp.Metadata) {
+		event := createEvent(data, metadata)
+		forwarder.Send(event)
+	}
+
+	server, err := tcp.New(cb, &config.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Input{
+		server:  server,
+		started: atomic.MakeBool(false),
+		outlet:  out,
+		config:  &config,
+		log:     logp.NewLogger("tcp input").With(config.Config.Host),
+	}, nil
+}
+
+// Run start a TCP input
+func (p *Input) Run() {
+	if !p.started.Load() {
+		p.log.Info("Starting TCP input")
+		err := p.server.Start()
+		if err != nil {
+			p.log.Errorw("Error starting the TCP server", "error", err)
+		}
+		p.started.Swap(true)
+	}
+}
+
+// Stop stops TCP server
+func (p *Input) Stop() {
+	p.log.Info("Stopping TCP input")
+	defer p.outlet.Close()
+	defer p.started.Swap(false)
+	p.server.Stop()
+}
+
+// Wait stop the current server
+func (p *Input) Wait() {
+	p.Stop()
+}
+
+func createEvent(raw []byte, metadata tcp.Metadata) *util.Data {
+	data := util.NewData()
+	data.Event = beat.Event{
+		Timestamp: time.Now(),
+		Meta: common.MapStr{
+			"ip_address": metadata.IPAddress,
```

Comment:
I suggest we have this information also in each event under something like `source.ip` or similar?

Reply:
+1 for source.ip when following ECS, but currently this will clash with `source` which is the original read file. So using `source.ip` will need to wait for 7.0, maybe we can use `client.ip` ?

```yaml
      description: > source
        The file from which the line was read. This field contains the absolute path to the file.
        For example: `/var/log/system.log`.
```

Reply:
The collision is unfortunate. Under ECS `source` will be moving to `file.path`. I think we have three options

1. Put the remote IP and port into `source` (e.g. `source: 127.0.0.1:38849`)
1. Leave it under `@metadata`. The value will be there if someone really needs it (they can use the rename processor to get it stored in ES). Then in 7.0 we move this to `source.ip`. Because it's tucked away in `@metadata` the impact of renaming it in 7.0 is minimized.
1. Store it in `client.ip` and rename it to `source.ip` for 7.0. This adds more fields that need to be renamed for 7.0, but on the plus side the value is more visible if people want to use it.

I like number 1.

Reply:
1. make sense, I didn't see like that.

Reply:
This is effectively the `source` file for it.

Reply:
```json
{
  "@timestamp": "2018-04-03T20:32:55.908Z",
  "@metadata": {
    "beat": "filebeat",
    "type": "doc",
    "version": "7.0.0-alpha1"
  },
  "source": "127.0.0.1:61656",
  "prospector": {
    "type": "tcp"
  },
  "input": {
    "type": "tcp"
  },
  "beat": {
    "name": "sashimi",
    "hostname": "sashimi",
    "version": "7.0.0-alpha1"
  },
  "message": "heee"
}
```